src/system/ZAuthModule/Controller/AccountController.php

```diff
@@ -147,55 +150,88 @@ public function lostPasswordAction(Request $request)
     }
 
     /**
-     * @Route("/lost-password/code")
+     * @Route("/lost-password/reset")
      * @Template
      * @param Request $request
      * @return array
      */
-    public function confirmationCodeAction(Request $request)
+    public function lostPasswordResetAction(Request $request)
     {
+        $redirectToRoute = 'zikulausersmodule_account_menu';
+
         if ($this->get('zikula_users_module.current_user')->isLoggedIn()) {
-            return $this->redirectToRoute('zikulausersmodule_account_menu');
+            return $this->redirectToRoute($redirectToRoute);
+        }
+
+        if (!$request->query->has('id')) {
+            $this->addFlash('error', $this->__('Your request could not be processed due to missing arguments.'));
+
+            return $this->redirectToRoute($redirectToRoute);
+        }
+
+        $lostPasswordVerificationHelper = $this->get('zikula_zauth_module.helper.lost_password_verification_helper');
+        $requestDetails = [];
+
+        try {
+            $requestDetails = $lostPasswordVerificationHelper->decodeLostPasswordId($request->query->get('id'));
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->__('Your request could not be processed.') . ' ' . $e->getMessage());
+
+            return $this->redirectToRoute($redirectToRoute);
+        }
+
+        if ($requestDetails['userId'] == '' || $requestDetails['userName'] == '' || $requestDetails['emailAddress'] == '') {
+            $this->addFlash('error', $this->__('Your request could not be processed due to invalid arguments.'));
+
+            return $this->redirectToRoute($redirectToRoute);
+        }
+
+        $user = $this->get('zikula_users_module.user_repository')->findBy([
+            'uid' => $requestDetails['userId'],
+            'uname' => $requestDetails['userName'],
+            'email' => $requestDetails['emailAddress']
+        ]);
```

Comment:
that's simply not true. The previous code only searched based on email. Such a search may possibly return several rows because in the past, the email has not been required to be unique. You are searching based on multiple params including the uid, which must be unique, therefore it will only return one row and using the `find` method is appropriate.